analyzers/tests/SonarAnalyzer.UnitTest/TestCases/LockedFieldShouldBePrivateAndReadonly.CSharp9.cs

```diff
@@ -0,0 +1,23 @@
+﻿class Test
+{
+    readonly ARecord readonlyField = new();
+    ARecord readWriteField = new();
+
+    static readonly ARecord staticReadonlyField = new();
+    static ARecord staticReadWriteField = new();
+
+    void OnAFieldOfTypeRecord()
+    {
+        lock (readonlyField) { }
+        lock (readWriteField) { }       // Noncompliant
+        lock (staticReadonlyField) { }
+        lock (staticReadWriteField) { } // Noncompliant
+    }
+
+    void OnANewRecordInstance()
+    {
+        lock (new ARecord()) { }  // Noncompliant
+    }
+
+    record ARecord();
+}
```

Comment:
Maybe it worths adding a small example also for top-level statements?

Reply:
Indeed! I forgot it, that's a good one. 
All keeps working because "top-level statements" is just syntactic sugar of an implicitly defined `Program` class.
It is also an interesting scenario because classes that syntactically appear top level are actually nested, and classes that syntactically appear 1st level nested are actually 2nd level. I have added those cases to the test as well.

Reply:
@antonioaversa  Did you add the top-level example? I cannot see.
Thinking about it - the rule should not raise an issue in top-level statements - as you cannot have private readonly fields there -  But it's still an valid scenario.

But this might be a good idea for a hardening ticket?

Reply:
Yes, it is in a dedicated file analyzers/tests/SonarAnalyzer.UnitTest/TestCases/LockedFieldShouldBePrivateAndReadonly.CSharp9.TopLevelStatements.cs

"You cannot have private readonly fields there": well, you can actually :-)

Because these "top-level statements" are just syntactic sugar for a `Program` class, you can add the fields by defining a partial class `Program`. The following code compiles:
```cs
void Method()
{
    lock (lockObj) // This lockObj is in scope
    {

    }
}

partial class Program
{
    private static readonly object lockObj = new();
}
```

So the rule makes sense, in its current form, for top-level statements too.

Reply:
Added a test with this scenario.

Reply:
After offline discussion with @antonioaversa and comments from @pavel-mikula-sonarsource locking in a top-level statement local variable is an edge case scenario that we don;t mind for now in this rule.

The rest of the UTs in the top-level sta. file seem to be covered by the already existing UTs - if it's not the case @antonioaversa maybe you could leave them - but IMO the should not be needed.

Reply:
Removed.

The rest of the unit tests covered the case of a nested class at IL level which is syntactically top-level (i.e. it does not have an ancestor of `ClassDeclarationSyntax`). That only happens in top-level statements, and that is what made the test different from the other tests.

However, because we have now narrowed the scope of S2445 for C#, to not check accessibility, we don't have different messages for the cases where the lock is being acquired on a field of the same class vs another class. Therefore, it does not make sense to explore scenarios involving class-checking, and the tests could be removed.